src/openssl/pkcs5.cr

```diff
@@ -8,4 +8,15 @@ module OpenSSL::PKCS5
     end
     buffer
   end
+
+  {% if compare_versions(LibSSL::OPENSSL_VERSION, "1.0.0") >= 0 %}
+    def self.pbkdf2_hmac(secret, salt, iterations = 2**16, algorithm : OpenSSL::Algorithm = OpenSSL::Algorithm::SHA1, key_size = 64) : Bytes
```

Comment:
In my opinion this method should always exist, but raise in case it's not available because of the OpenSSL version. Please change it. Thank you!

Reply:
Good catch!

Reply:
To clarify, raising at compile time (e.g `method X not supported with OpenSSL version X.X`) instead of having an unknown method.

Reply:
No, this should raise at runtime.

Reply:
Why? The program is compiled with an OpenSSL version that doesn't support the aglorithm.

Reply:
Imagine this function is used conditionally, with a condition at runtime. Now imagine this is in a shard. I link that shard and want to use it, configuring it to not use this function: I can't, I won't be able to compile the program, even if at runtime I know I won't use the function. It's better to raise at runtime (I can change a config and avoid the exception).

Reply:
You'll be able to compile with a macro condition at compile time. What if you think the function won't be used but it does in a particular situation? This is a case that could be dealed before runtime, and your application will be more robust because this error won't exist.

Reply:
Right, then you need to use a macro condition everywhere, which is tedious.

This has already been discussed in the past, I can't remember the conclusion.

Reply:
Methods shouldn't just appear and disappear based on environmental reasons. It makes documentation difficult and breaks basic expectations. This has already been decided, i can't find this issue but it was discussed in length.

Reply:
That's fair. Probably this can be added somewhere (Wiki?) as contribution best practices.

Reply:
Good idea. I'll later add it to the FAQ, or Contributing.

Reply:
I'm understanding that I should accept the suggestion of @asterite: remove the macro conditional and throw an error at runtime.

Am I right?

Reply:
Is it **really** needed to support OpenSSL versions older than 1.0.0? Even CentOS 6 isn't on this version (seems to be 1.0.1e).
1.0.0 was released in [2010](https://en.wikipedia.org/wiki/OpenSSL#Major_version_releases)

Reply:
@mniak yes

@j8r thats for another issue